### PR TITLE
k8swatch: fix a race condition in pod change events

### PR DESCRIPTION
**Pull request by nicks**
Hello @landism, @maiamcc,

Please review the following commits I made in branch nicks/ch4899/update:

75cbac9610e01c610b205d1b2c03faa296ef99dd (2020-02-27 17:41:14 -0500)
k8swatch: fix a race condition in pod change events
Sometimes pod events would show up in the wrong order due to goroutine scheduling

Code review reminders, by giving a LGTM you attest that:

* Commits are adequately tested
* Code is easy to understand and conforms to style guides
* Incomplete code is marked with TODOs
* Code is suitably instrumented with logging and metrics

### CODE REVIEW

**Comment by nicks** [comment]
i think this is what's causing some of the weird integration test failures from time to time

**Comment by landism** [comment]
IIUC, this fixes the potential for dispatching pod events out of order, but it replaces out-of-order pod events with duplicates of the latest pod event? This is mostly fine for the purposes of ensuring that the engine state has the latest pod state, but seems bad when we care about intermediate state, e.g., maybe we skip logging one of the stages of pod rollout.

Would it be unreasonable to fix the race condition by dropping the `go` when calling `dispatchPodChange`? (or, if we don't want to block goroutines, having a chan of pod events that we dispatch in order)

**Comment by nicks** [comment]
Ya, this is the same API guarantee that the informers make:
https://godoc.org/k8s.io/client-go/tools/cache#ResourceEventHandler

> OnUpdate is called when an object is modified. Note that oldObj is the
    last known state of the object-- it is possible that several changes
    were combined together, so you can't use this to see every single
    change. OnUpdate is also called when a re-list happens, and it will
    get called even if nothing changed. This is useful for periodically
    evaluating or syncing something.

in other words, the underlying informer APIs also dispatch duplicate events, and also expect that intermediate states will get collapsed together. So I don't think this is something we can prevent.
